src/transforms/hexbin.js

```diff
@@ -18,45 +17,45 @@ export function hexbin(outputs = {fill: "count"}, options = {}) {
 
 // TODO filter e.g. to show empty hexbins?
 // TODO disallow x, x1, x2, y, y1, y2 reducers?
-function hexbinn(outputs, {binWidth = 20, fill, stroke, z, ...options}) {
+function hexbinn(outputs, {binWidth = 20, z, fill, stroke, ...options}) {
   binWidth = +binWidth;
-  const [GZ, setGZ] = maybeColumn(z);
   const [vfill] = maybeColorChannel(fill);
   const [vstroke] = maybeColorChannel(stroke);
-  const [GF = fill, setGF] = maybeColumn(vfill);
-  const [GS = stroke, setGS] = maybeColumn(vstroke);
-  outputs = maybeOutputs({
-    ...setGF && {fill: "first"},
-    ...setGS && {stroke: "first"},
-    ...outputs
-  }, {fill, stroke, ...options});
+  outputs = maybeOutputs(outputs, {z, fill, stroke, ...options});
   return {
     symbol: "hexagon",
     ...!hasOutput(outputs, "r") && {r: binWidth / 2},
-    ...!setGF && {fill},
-    ...((hasOutput(outputs, "fill") || setGF) && stroke === undefined) ? {stroke: "none"} : {stroke},
+    ...!hasOutput(outputs, "fill") && {fill},
+    ...((hasOutput(outputs, "fill") || vstroke != null) && stroke === undefined) ? {stroke: "none"} : {stroke},
     ...initialize(options, function(data, facets, {x: X, y: Y}, scales) {
-      if (setGF) setGF(valueof(data, vfill));
-      if (setGS) setGS(valueof(data, vstroke));
-      if (setGZ) setGZ(valueof(data, z));
-      for (const o of outputs) o.initialize(data);
       if (X === undefined) throw new Error("missing channel: x");
       if (Y === undefined) throw new Error("missing channel: y");
       const x = X.scale !== undefined ? scales[X.scale] : identity.transform;
       const y = Y.scale !== undefined ? scales[Y.scale] : identity.transform;
       X = X.value.map(x);
       Y = Y.value.map(y);
-      const F = setGF && GF.transform();
-      const S = setGS && GS.transform();
-      const Z = setGZ ? GZ.transform() : (F || S);
+      const Z = valueof(data, z);
+      const F = valueof(data, vfill);
+      const S = valueof(data, vstroke);
+      const G = maybeSubgroup(outputs, Z, F, S);
+      if (Z && !outputs.find(r => r.name === "z")) {
+        outputs.push(...maybeOutputs({z: "first"}, {z: Z}));
```

Comment:
This is mutating `outputs` which is declared up above (in the `hexbinn` scope). This is bad because it means if the initializer is invoked more than once, it will behave differently. Generally speaking we don’t invoke an initializer more than once (since Plot.plot is often invoked once and then everything is thrown away), but still, we should try to write the transforms to be stateless. Invoking them shouldn’t mutate the internal state of the transform.

Reply:
Also, I’m very confused as to why the logic here is so different than what the bin and group transforms do… Does it have something to do with the fact that initializers run after channels have been constructed?

Reply:
I came up with an approach I like in 2dc40c09aeba3238572b153b8c78c375c2b1fb74.